ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/ParseReportingUrl.java

```diff
@@ -0,0 +1,302 @@
+
+          // TODO: Validate required query params
+
+          queryParams.put("ip",
+              createIpParam(payload.get(Attribute.NORMALIZED_COUNTRY_CODE).asText()));
+
+          try {
+            queryParams.put("ua", createUserAgentParam(attributes.get(Attribute.USER_AGENT_OS),
+                payload.get(Attribute.VERSION).asText()));
+          } catch (UnsupportedEncodingException e) {
+            throw new UncheckedIOException(e);
+          }
+
+          // Generate query string from map
+          String queryString = queryParams.entrySet().stream().map(
+              entry -> entry.getKey() + (entry.getValue() == null ? "" : "=" + entry.getValue()))
+              .collect(Collectors.joining("&"));
+
+          try {
+            reportingUrl = new URL(urlObj.getProtocol(), urlObj.getHost(),
+                urlObj.getPath() + "?" + queryString).toString();
+          } catch (MalformedURLException e) {
+            throw new InvalidUrlException(
+                "Could not parse reporting with query string: " + queryString, e);
+          }
+
+          attributes.put(Attribute.REPORTING_URL, reportingUrl);
+          payload.put(Attribute.REPORTING_URL, reportingUrl);
+
+          return new PubsubMessage(Json.asBytes(payload), attributes);
+        }).exceptionsInto(TypeDescriptor.of(PubsubMessage.class))
+            .exceptionsVia((ExceptionElement<PubsubMessage> ee) -> {
+              try {
+                throw ee.exception();
+              } catch (UncheckedIOException | IllegalArgumentException | InvalidUrlException e) {
+                return FailureMessage.of(ParseReportingUrl.class.getSimpleName(), ee.element(),
+                    ee.exception());
+              }
+            }));
+  }
+
+  @VisibleForTesting
+  boolean isUrlValid(URL url, String doctype) {
+    Set<String> allowedUrls;
+
+    if (doctype.endsWith("-click")) {
+      allowedUrls = singletonAllowedClickUrls;
+    } else if (doctype.endsWith("-impression")) {
+      allowedUrls = singletonAllowedImpressionUrls;
+    } else {
+      throw new IllegalArgumentException("Invalid doctype: " + doctype);
+    }
+
+    if (allowedUrls.contains(url.getHost())) {
+      return true;
+    }
+
+    // check for subdomains (e.g. allow mozilla.test.com but not mozillatest.com)
+    return allowedUrls.stream().map(allowedUrl -> "." + allowedUrl)
+        .anyMatch(url.getHost()::endsWith);
+  }
+
+  @VisibleForTesting
+  String createIpParam(String countryCode) {
+    String ipParam = singletonCountryToIpMapping.getOrDefault(countryCode,
+        singletonCountryToIpMapping.get(DEFAULT_COUNTRY));
+    if (ipParam == null) {
+      throw new IllegalArgumentException("Could not get ip value: Unrecognized country "
+          + "and missing default value: " + Attribute.NORMALIZED_COUNTRY_CODE);
+    }
+    return ipParam;
+  }
+
+  @VisibleForTesting
+  String createUserAgentParam(String os, String clientVersion) throws UnsupportedEncodingException {
```

Comment:
```suggestion
  String createUserAgentParam(String userAgentOs, String clientVersion) throws UnsupportedEncodingException {
```

As I was reading this, I was wondering, "why don't we use normalized_os?" because I forgot we only have user agent os. It would be nice to always refer to this non-normalized value as userAgentOs to make that more explicit.